content/blog/mozilla/2018-07-03.md

```diff
@@ -0,0 +1,217 @@
+Title: Making calls to WebAssembly fast and implementing anyref
+Date: 2018-07-02 20:00
+Author: Benjamin Bouvier
+Tags: mozilla
+Slug: mozilla-2018-h1
+Lang: en
+
+Since this is the end of the first half-year, I think it is a good time to
+reflect and show some work I've been doing over the last few months, apart from
+the regular batch of random issues, security bugs, reviews and the fixing of 24
+bugs found by our [fuzzers](https://en.wikipedia.org/wiki/Fuzzing).
+
+## Bug 1319203: Make JS to WebAssembly calls *blazingly* fast
+
+If we want more WebAssembly (wasm) adoption, there shouldn't be a big costly
+barrier between the two universes. That is, calls from one world to the other
+should be fast. For a very long time, calls from JS to asm.js/WebAssembly have
+been quite slow in Firefox. In fact, we didn't optimize them at all. For ease
+and speed of implementation at the time, asm.js call activations (data
+structures recording information about the function being currently called in
+the VM) were very different from the JS ones, implying some other huge
+structural differences, like the capability to reconstruct call stack
+information used by `Error()` stack frames, or just tracing the stack for
+garbage collection purposes. After many refactorings and low-level changes that
+took a great bunch of last year for me, Spidermonkey was ripe for an
+optimization.
+
+To give some context here, when we call from JS to asm.js/wasm, and the JS
```

Comment:
it might be easier for folks to read if this was broken out into several sentences:

"The critical step here is JIT compilation. JIT compilation means that the code is compiled to machine code by the just-in-time compiler, IonMonkey. When JS calls into asm.js/wasm, this JS caller *hasn't* been JIT-compiled. As a result, the call passes through C++, does a bunch of work, and then this C++ calls into a piece of glue code directly written in assembly: the *interpreter entry stub*. This stub is quite small: it copies out the C++ arguments into argument positions the wasm function requires, sets up some small machine state, calls into the function, performs error checking, and finally returns to the C++ caller."

Reply:
this paragraph would do very well with a diagram. happy to make one for you if you are interested!

Reply:
PR with diagrams #8

Reply:
wow, I realize my explanation here was kind of incorrect: before the patches, we'd have to go through the interpreter entry *even for* JS functions that were JIT-compiled (meaning, they'd go from jit code to C++ first, then C++ to wasm, etc.). The diagrams remain valid, just the wording needs to change. I'll try to write short sentences here. Thanks!